### PR TITLE
Replace unsafe functions with the safer options

### PR DESCRIPTION
**Pull request by basasairohan**
### Description

Currently in `check_server_collation_name` function in `guc.c` file we are using `strcpy` which may cause buffer overflow. This change replaces it with `strncpy` to avoid buffer overflow

Task: BABEL-3581
Signed-off-by: Sai Rohan Basa [bsrohan@amazon.com](mailto:bsrohan@amazon.com)
 
### Issues Resolved

[List any issues this PR will resolve]

### Test Scenarios Covered ###
* **Use case based -** We have tested the changes manually. But it is currently not possible to add automated tests

* **Boundary conditions -** N/A

* **Arbitrary inputs -** N/A

* **Negative test cases -** N/A

* **Minor version upgrade tests -** N/A

* **Major version upgrade tests -** N/A

* **Performance tests -** N/A

* **Tooling impact -** N/A

* **Client tests -** N/A

### Check List
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: contrib/babelfishpg_tsql/src/guc.c [hunk 1]
```diff
@@ -136,7 +136,8 @@ static bool check_server_collation_name(char **newval, void **extra, GucSource s
 			* Collation names are stored in lowercase into pg catalog (pg_collation).
 			*/
 		char *dupval = pstrdup(*newval);
-		strcpy(*newval, downcase_identifier(dupval, strlen(dupval), false, false));
+		Assert(sizeof(newval) >= sizeof(dupval));
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/guc.c]: this check has no value, it is checking for char * size

File: contrib/babelfishpg_tsql/src/guc.c [hunk 2]
```diff
@@ -136,7 +136,8 @@ static bool check_server_collation_name(char **newval, void **extra, GucSource s
 			* Collation names are stored in lowercase into pg catalog (pg_collation).
 			*/
 		char *dupval = pstrdup(*newval);
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/guc.c]: overall, there should be better way to this. We are creating unnecessary two copies just for down casing

File: contrib/babelfishpg_tsql/src/guc.c [hunk 3]
```diff
@@ -135,9 +135,7 @@ static bool check_server_collation_name(char **newval, void **extra, GucSource s
 			* We are storing value in lower case since
 			* Collation names are stored in lowercase into pg catalog (pg_collation).
 			*/
-		char *dupval = pstrdup(*newval);
-		strcpy(*newval, downcase_identifier(dupval, strlen(dupval), false, false));
-		pfree(dupval);
+		strncpy(*newval, downcase_identifier(*newval, strlen(*newval), false, false),strlen(*newval));
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/guc.c]: Instead of calling strlen twice, call it once and pass stored len value

**Review by surendravishnoi — COMMENTED** [review]
Please use new commit message template

File: contrib/babelfishpg_tsql/src/guc.c (hunk 3 above)
Comment (basasairohan) [reply contrib/babelfishpg_tsql/src/guc.c]: okay

**Review by Deepesh125 — DISMISSED** [review]

File: contrib/babelfishpg_tsql/src/guc.c [hunk 4]
```diff
@@ -135,9 +135,8 @@ static bool check_server_collation_name(char **newval, void **extra, GucSource s
 			* We are storing value in lower case since
 			* Collation names are stored in lowercase into pg catalog (pg_collation).
 			*/
-		char *dupval = pstrdup(*newval);
-		strcpy(*newval, downcase_identifier(dupval, strlen(dupval), false, false));
-		pfree(dupval);
+		int length = strlen(*newval);
+		strncpy(*newval, downcase_identifier(*newval, length, false, false),length);
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/guc.c]: space before length

**Review by surendravishnoi — COMMENTED** [review]
Please use commit template for PR commit message

**Review by surendravishnoi — APPROVED** [review]

**Pull request closed without merging**